test/Kestrel.FunctionalTests/Http2/ShutdownTests.cs

```diff
@@ -0,0 +1,197 @@
+    {
+        private static X509Certificate2 _x509Certificate2 = TestResources.GetTestCertificate();
+
+        private HttpClient Client { get; set; }
+        private List<Http2Frame> ReceivedFrames { get; } = new List<Http2Frame>();
+
+        public ShutdownTests()
+        {
+            if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+            {
+                // We don't want the default SocketsHttpHandler, it doesn't support HTTP/2 yet.
+                Client = new HttpClient(new WinHttpHandler()
+                {
+                    ServerCertificateValidationCallback = HttpClientHandler.DangerousAcceptAnyServerCertificateValidator
+                });
+            }
+        }
+
+        [ConditionalFact]
+        public async Task GracefulShutdownWaitsForRequestsToFinish()
+        {
+            var requestStarted = new ManualResetEventSlim();
+            var requestUnblocked = new ManualResetEventSlim();
+            using (var server = new TestServer(async context =>
+            {
+                requestStarted.Set();
+                Assert.True(requestUnblocked.Wait(5000), "request timeout");
+                await context.Response.WriteAsync("hello world " + context.Request.Protocol);
+            }, new TestServiceContext(LoggerFactory),
+            kestrelOptions =>
+            {
+                kestrelOptions.Listen(IPAddress.Loopback, 0, listenOptions =>
+                {
+                    listenOptions.Protocols = HttpProtocols.Http2;
+                    listenOptions.UseHttps(_x509Certificate2);
+                });
+            }))
+            {
+                // Setup a listening stream
+                var connection = server.CreateConnection();
+                var sslStream = new SslStream(connection.Stream);
+                await sslStream.AuthenticateAsClientAsync(new SslClientAuthenticationOptions()
+                {
+                    TargetHost = "localhost",
+                    RemoteCertificateValidationCallback = (_, __, ___, ____) => true,
+                    ApplicationProtocols = new List<SslApplicationProtocol>() { SslApplicationProtocol.Http2, SslApplicationProtocol.Http11 },
+                    EnabledSslProtocols = SslProtocols.Tls12,
+                }, CancellationToken.None);
+                var reader = PipeReaderFactory.CreateFromStream(PipeOptions.Default, sslStream, CancellationToken.None);
+
+                var requestTask = Client.GetStringAsync($"https://localhost:{server.Port}/");
+                Assert.False(requestTask.IsCompleted);
+
+                Assert.True(requestStarted.Wait(5000), "timeout");
+
+                var stopTask = server.StopAsync();
+
+                // Unblock the request
+                requestUnblocked.Set();
+
+                Assert.Equal("hello world HTTP/2", await requestTask);
+                Assert.Equal(stopTask, await Task.WhenAny(stopTask, Task.Delay(10000)));
+
+                await ReceiveFramesAsync(reader);
```

Comment:
? How are you intercepting a raw frame here? Shouldn't the server have read that?

Reply:
What do you mean the server should have read that? I open a stream and wait for the connection to send GOAWAYs as part of its lifecycle.